File: app/src/main/java/com/peterlaurence/trekadvisor/core/sensors/OrientationSensor.java
package com.peterlaurence.trekadvisor.core.sensors;

import android.app.Activity;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.view.Surface;

import java.lang.ref.WeakReference;

/**
 * A facility object that retrieves the orientation of the phone.
 *
 * @author peterLaurence on 09/04/16.
 */
public class OrientationSensor implements SensorEventListener {

    private WeakReference<OrientationListener> mOrientationListenerWeakReference;

    /* object internals */
    private boolean mStarted;
    private Activity mActivity;
    private float[] orientationValues;
    private float[] rMat;


    public OrientationSensor(Activity context) {
        mActivity = context;
        orientationValues = new float[3];
        rMat = new float[9];

        start();
    }

    /**
     * Set a listener so that it will be called everytime the azimuth changes
     *
     * @param listener a {@link OrientationListener} object
     */
    public void setOrientationListener(OrientationListener listener) {
        mOrientationListenerWeakReference = new WeakReference<>(listener);
        if (mStarted) {
            listener.onOrientationEnable();
        } else {
            listener.onOrientationDisable();
        }
    }

    public boolean toggleOrientation() {
        if (mStarted) {
            stop();
        } else {
            start();
        }
        return mStarted;
    }

    public boolean isStarted() {
        return mStarted;
    }

    public void start() {
        mStarted = true;

        /* First subscribe to orientation events */
        SensorManager sensorManager = (SensorManager) mActivity.getSystemService(Context.SENSOR_SERVICE);
        Sensor mSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
        sensorManager.registerListener(this, mSensor, 100);

        /* Then update the listener (view) */
        if (mOrientationListenerWeakReference != null) {
            OrientationListener listener = mOrientationListenerWeakReference.get();
            if (listener != null) {
                listener.onOrientationEnable();
            }
        }
    }

    public void stop() {
        mStarted = false;

        /* First update the listener (view) */
        if (mOrientationListenerWeakReference != null) {
            OrientationListener listener = mOrientationListenerWeakReference.get();
            if (listener != null) {
                listener.onOrientationDisable();
            }
        }

        /* Then unsubscribe */
        SensorManager sensorManager = (SensorManager) mActivity.getSystemService(Context.SENSOR_SERVICE);
        sensorManager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() == Sensor.TYPE_ROTATION_VECTOR) {
            /* Calculate the rotation matrix */
            SensorManager.getRotationMatrixFromVector(rMat, event.values);

            int screenRotation = mActivity.getWindowManager().getDefaultDisplay().getRotation();
            int fix = 0;
            switch (screenRotation) {
                case Surface.ROTATION_90:
                    fix = 90;
                    break;

                case Surface.ROTATION_180:
                    fix = 180;
                    break;

                case Surface.ROTATION_270:
                    fix = 270;
                    break;

                default:
                    break;
            }

            /* Get the azimuth value (orientation[0]) in degree */
            int mAzimuth = (int) (Math.toDegrees(SensorManager.getOrientation(rMat, orientationValues)[0]) + 360 + fix) % 360;

            /* Call the listener */
            if (mOrientationListenerWeakReference != null) {
                OrientationListener listener = mOrientationListenerWeakReference.get();
                if (listener != null) {
                    listener.onOrientation(mAzimuth);
                }
            }
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }

    public interface OrientationListener {
        void onOrientation(int azimuth);

        void onOrientationEnable();

        void onOrientationDisable();
    }
}

File: app/src/main/java/com/peterlaurence/trekme/core/sensors/OrientationSensor.kt
package com.peterlaurence.trekme.core.sensors

import android.app.Activity
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.view.Surface
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlin.math.abs

class OrientationSensor(val activity: Activity) : SensorEventListener {
    private val sensorManager: SensorManager = activity.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val rotationVectorReading = FloatArray(3)

    private val rotationMatrix = FloatArray(9)
    private val orientationAngles = FloatArray(3)

    var isStarted: Boolean = false
        private set

    fun getAzimuthFlow(): Flow<Float> = flow {
        while (isStarted) {
            delay(1)
            updateOrientation()
            val screenRotation: Int = activity.windowManager.defaultDisplay.rotation
            var fix = 0
            when (screenRotation) {
                Surface.ROTATION_90 -> fix = 90
                Surface.ROTATION_180 -> fix = 180
                Surface.ROTATION_270 -> fix = 270
            }

            /* Get the azimuth value (orientation[0]) in degree */
            val azimuth = (Math.toDegrees(orientationAngles[0].toDouble()) + 360 + fix).toFloat() % 360

            emit(azimuth)
        }
    }.distinctUntilChanged { old, new -> abs(old - new) < 0.1 }.flowOn(Dispatchers.Default)

    fun start() {
        sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)?.also { sensor ->
            sensorManager.registerListener(
                    this,
                    sensor,
                    SensorManager.SENSOR_DELAY_FASTEST
            )
        }
        isStarted = true
    }

    fun stop() {
        sensorManager.unregisterListener(this)
        isStarted = false
    }

    fun toggle(): Boolean {
        if (isStarted) stop() else start()
        return isStarted
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (event.sensor.type == Sensor.TYPE_ROTATION_VECTOR) {
            System.arraycopy(event.values, 0, rotationVectorReading, 0, rotationVectorReading.size)
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
    }

    private fun updateOrientation() {
        SensorManager.getRotationMatrixFromVector(rotationMatrix, rotationVectorReading)

        /* Get the azimuth value (orientation[0]) in degree */
        SensorManager.getOrientation(rotationMatrix, orientationAngles)
    }
}
